test/bin/generate_common_versions.py

```diff
@@ -0,0 +1,253 @@
+
+    ``provides_pkg`` checks if the repository provides the package specified by `pkg`.
+    """
+    if repo.startswith("https"):
+        try:
+            temp = f"this,{repo}"
+            _ = subprocess.run(
+                    ['dnf', 'repoquery', pkg, '--quiet',
+                     '--disablerepo', '*', '--repofrompath', temp],
+                    capture_output=True,
+                    check=True
+                )
+
+            return True
+
+        except subprocess.CalledProcessError:
+            return False
+    else:
+        try:
+            _ = subprocess.run(
+                    ['dnf', 'repoquery', pkg, '--quiet',
+                     '--repo', repo],
+                    capture_output=True,
+                    check=True
+            )
+
+            return True
+
+        except subprocess.CalledProcessError:
+            return False
+
+
+def dnf_repo_is_enabled(repo_name):
+    """
+    :param repo_name: the name of the repository to check
+
+    ``dnf_repo_is_enabled`` checks if the repository is enabled on this system.
+    """
+    try:
+        result = subprocess.run(
+            ["dnf", "repolist"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.DEVNULL,
+            text=True,
+            check=True
+        )
+        return repo_name in result.stdout
+    except subprocess.CalledProcessError:
+        return False
+
+
+def repo_exists(repo_url):
+    """
+    :param repo_url: the URL of a repository
+
+    ``repo_exists`` checks if a URL points to a valid repository.
+    """
+    url = repo_url + "/repodata/repomd.xml"
+    r = requests.get(url)
+    if r.status_code == 404:
+        return False
+    else:
+        return True
+
+
+def get_sub_repo(minor):
+    """
+    :param minor: the minor version, e.g. 19 for 4.19
+
+    ``get_sub_repo`` returns the name of the subscription repository for the specified
+    minor version if the repository provides the microshift package, otherwise returns None.
+    """
+    try:
+        repo = f"rhocp-4.{minor}-for-rhel-9-{ARCH}-rpms"
+        _ = subprocess.run(
+                ['sudo', 'dnf', 'repoquery', 'microshift', '--quiet',
+                 '--queryformat', '%{version}-%{release}', '--repo', repo],
+                capture_output=True,
+                text=True,
+                check=True
+            )
+
+        return repo
+
+    except subprocess.CalledProcessError:
+        return None
+
+
+def get_release_repo(minor):
+    """
+    :param minor: the minor version, e.g. 19 for 4.19
+
+    ``get_release_repo`` returns the repository for the specified minor version.
+    It looks for the 'rhocp' stream, release candidate and engineering candidate,
+    in that order. If none of these repositories are available, returns empty
+    double quotes (to be used as empty string in bash).
+    """
+    repo = get_sub_repo(minor)
+    if repo is not None and provides_pkg(repo, "microshift"):
```

Comment:
I think it's a bit cyclic.
In first call we're checking if rhocp for `4.$minor` contains microshift, then use `provides_pkg(repo, "microshift")`. after checking if `repo is not None` we already know that it's there.